### PR TITLE
Fix: App icon in launch (aka splash) screen should not move (up) abruptly #1897

### LINKED ISSUE #1897: App icon in launch (aka splash) screen should not move (up) abruptly
During a cold launch, this happens:

1. the launch screen (which includes the app icon) appears
2. SplashView(Controller) appears (probably due to `ProtectionCoordinator`) which includes the app icon too.
3. SplashViewController is hidden

The app icon in (1) and (2)'s y position is slightly different so it looks like the logo moved up abruptly. The icon should not move at all.

Need to verify the fix works on 5s, iPhone X and a bigger than 5s phone with no notch.

### PR DESCRIPTION
**Pull request by hadwanihardik**
Fixes #1897

### CODE REVIEW

File: AlphaWallet/Extensions/UIDevice.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//  AlphaWallet
+//
+//  Created by Hardik on 6/17/20.
+//
+
+import Foundation
+import UIKit
+
+extension UIDevice {
+    var hasNotch: Bool {
```
Comment (hboon) [inline AlphaWallet/Extensions/UIDevice.swift:12]: Can you fix it without add `hasNotch`? This is very brittle.

File: AlphaWallet/Protection/Coordinators/ProtectionCoordinator.swift [hunk 2]
```diff
@@ -21,7 +21,7 @@ class ProtectionCoordinator: Coordinator {
 	}
 
 	func didFinishLaunchingWithOptions() {
-		splashCoordinator.start()
+		//splashCoordinator.start()
```
Comment (hboon) [inline AlphaWallet/Protection/Coordinators/ProtectionCoordinator.swift:24]: Commented out code should generally not be checked in, they should be deleted.

What is the rationale for removing this though?

File: AlphaWallet/Protection/Coordinators/SplashCoordinator.swift [hunk 3]
```diff
@@ -9,13 +9,16 @@ class SplashCoordinator: Coordinator {
 
     init(window: UIWindow) {
         self.window = window
+        window.backgroundColor =  .white;
```
Comment (hboon) [inline AlphaWallet/Protection/Coordinators/SplashCoordinator.swift:12]: What is this for?

File: AlphaWallet/Protection/Coordinators/SplashCoordinator.swift [hunk 4]
```diff
@@ -9,13 +9,16 @@ class SplashCoordinator: Coordinator {
...
     }
 
     func start() {
         window.rootViewController = SplashViewController()
         window.isHidden = false
+        window.rootViewController = SplashViewController()
```
Comment (hboon) [inline AlphaWallet/Protection/Coordinators/SplashCoordinator.swift:18]: Why is it being created and assigned twice?

File: AlphaWallet/Protection/ViewControllers/SplashViewController.swift [hunk 5]
```diff
@@ -8,14 +8,29 @@ class SplashViewController: UIViewController {
...
-            splashView.topAnchor.constraint(equalTo: view.topAnchor),
-            splashView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
-            splashView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
-            splashView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
-        ])
+        splashView.backgroundColor = UIColor.red; splashView.translatesAutoresizingMaskIntoConstraints = false
+        splashView.frame =  UIScreen.main.bounds
+       view.addSubview(splashView)
+       if UIDevice.current.hasNotch {
+        if #available(iOS 11, *) {
```
Comment (hboon) [inline AlphaWallet/Protection/ViewControllers/SplashViewController.swift:15]: * We should always follow the existing indentation style in the file (or if that file is really bad, fix the file).
* The app doesn't support < iOS 11, so the check is not needed

File: AlphaWallet/Protection/ViewControllers/SplashViewController.swift [hunk 6]
```diff
@@ -8,14 +8,29 @@ class SplashViewController: UIViewController {
...
+              let guide = view.safeAreaLayoutGuide
+              NSLayoutConstraint.activate([
+                splashView.topAnchor.constraint(equalToSystemSpacingBelow: guide.topAnchor, multiplier: 1.0),
+                splashView.bottomAnchor.constraint(equalToSystemSpacingBelow: guide.bottomAnchor, multiplier: 1.0)
+               ])
+        }
+        } else {
+           let standardSpacing: CGFloat = 8.0
+           NSLayoutConstraint.activate([
+           splashView.topAnchor.constraint(equalTo: view.bottomAnchor, constant: standardSpacing),
```
Comment (hboon) [inline AlphaWallet/Protection/ViewControllers/SplashViewController.swift:25]: Indentation, again

File: AlphaWallet/Protection/ViewControllers/SplashViewController.swift [hunk 7]
```diff
@@ -8,14 +8,29 @@ class SplashViewController: UIViewController {
...
+           splashView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: standardSpacing)
+           ])
+        }
+        splashView.layoutSubviews();
+
+    }
+    override func viewDidLayoutSubviews() {
```
Comment (hboon) [inline AlphaWallet/Protection/ViewControllers/SplashViewController.swift:32]: If it's empty, we should remove it

File: AlphaWallet/Protection/Views/SplashView.swift [hunk 8]
```diff
@@ -3,16 +3,31 @@
...
         addSubview(logoImageView)
-        NSLayoutConstraint.activate([
-            logoImageView.centerXAnchor.constraint(equalTo: centerXAnchor),
-            logoImageView.centerYAnchor.constraint(equalTo: centerYAnchor),
-        ])
+        self.layoutSubviews()
+    }
+    override func layoutSubviews() {
+        if UIDevice.current.hasNotch {
+                   let guide = safeAreaLayoutGuide
```
Comment (hboon) [inline AlphaWallet/Protection/Views/SplashView.swift:18]: Indentation. Ca you fix the indentation for the entire commit?

File: Podfile.lock [hunk 9]
```diff
@@ -252,4 +252,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 4d513ea9af5a9550a53c8b47493983d196f33d13
 
-COCOAPODS: 1.8.4
+COCOAPODS: 1.9.1
```
Comment (hboon) [inline Podfile.lock:255]: Generally, we try not to include any unnecessary changes in a PR

File: AlphaWallet/Protection/Coordinators/ProtectionCoordinator.swift (hunk 2 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/Coordinators/ProtectionCoordinator.swift:24]: Actually this part to code is adding splash view on window even before app is loaded so its not getting proper frame coordinates. because of that when we open app after creating wallet it shows black screen with logo with layout issue.
`splashCoordinator.start()`
As its already called on `applicationDidBecomeActive` we can remove this line of code in 
`didFinishLaunchingWithOptions`.

File: AlphaWallet/Protection/Coordinators/SplashCoordinator.swift (hunk 3 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/Coordinators/SplashCoordinator.swift:12]: This was making background color of splash view controller white as on app launch when its loaded on window object it was showing black  screen with logo after wallet creation

File: AlphaWallet/Protection/Coordinators/SplashCoordinator.swift (hunk 4 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/Coordinators/SplashCoordinator.swift:18]: it was by mistake removing it. one part of it was commented.

File: AlphaWallet/Protection/ViewControllers/SplashViewController.swift (hunk 5 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/ViewControllers/SplashViewController.swift:15]: ok sure I am making note of this.

File: AlphaWallet/Protection/ViewControllers/SplashViewController.swift (hunk 6 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/ViewControllers/SplashViewController.swift:25]: ok

File: AlphaWallet/Protection/ViewControllers/SplashViewController.swift (hunk 7 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/ViewControllers/SplashViewController.swift:32]: ok

File: AlphaWallet/Protection/Views/SplashView.swift (hunk 8 above)
Comment (hadwanihardik) [reply AlphaWallet/Protection/Views/SplashView.swift:18]: ok

File: Podfile.lock (hunk 9 above)
Comment (hadwanihardik) [reply Podfile.lock:255]: ok

**Comment by hadwanihardik** [comment]
Made Changes as per suggestion by boon.

1. Made indentation proper for all files in commit.
2. Removed unnecessary commented code and spaces.
3. Removed condition for check ios version lower than ios 11.
4. Removed code for checking notch display, and removed extension file related to same.
5. Removed blank methods.

**Comment by hboon** [comment]
Thanks. Would you help to rebase master and squash your changes into a single commit, and I’ll check again?

**Comment by hadwanihardik** [comment]
> Thanks. Would you help to rebase master and squash your changes into a single commit, and I’ll check again?

Hey I did combined changes in single commit.

**Comment by hboon** [comment]
It seems like you might have accidentally modified the 2nd commit in this PR. Can you compare to master? Your PR branch should basically be the tip of master + 1 commit

**Comment by hadwanihardik** [comment]
> It seems like you might have accidentally modified the 2nd commit in this PR. Can you compare to master? Your PR branch should basically be the tip of master + 1 commit

Hey I did it by mistake, first commit is not for this PR. How can i remove it from PR?

**Comment by hboon** [comment]
Rebase

**Comment by hadwanihardik** [comment]
> Rebase

I did not understood. I tried to rebase to my commit but it did not worked

**Comment by hboon** [comment]
Could you have accidentally squashed your commit with the commit at the tip of the master?

**Comment by hadwanihardik** [comment]
> Could you have accidentally squashed your commit with the commit at the tip of the master?

I don't have idea. if u can guide me?

**Comment by hadwanihardik** [comment]
> > Could you have accidentally squashed your commit with the commit at the tip of the master?
> 
> I don't have idea. if u can guide me?

may I generate a new PR with new branch?

**Comment by hboon** [comment]
> may I generate a new PR with new branch?

Ok

**Comment by hboon** [comment]
Superseded by #1966
